### ISSUE TITLE
High level API的一些使用问题

### ISSUE THREAD

**Issue by qingqing01** [PaddlePaddle/Paddle#11246]
背景是要升级一些models里模型，遇到若干问题：

1.  如何使得test_program和train_program不同？
    - 问题：类似以下定义，如何只给test_program里加一些op，train_program里不加？
       ```python
        loss = fluid.layers.ssd_loss(locs, confs, gt_box, gt_label, box,box_var)
        loss = fluid.layers.reduce_sum(loss)
        test_program = fluid.default_main_program().clone(for_test=True)
        with fluid.program_guard(test_program):
            map_eval = fluid.evaluator.DetectionMAP(...)
       ```
    - 当前内部实现问题：只从train_program里clone出来test_program。(**也没有加for_test=True，这样对有BatchNorm、Dropout的网络在测试时的行为是正确的吗？**)
        [self.test_program = self.train_program.clone()](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/trainer.py#L118)

2. 如何Load部分参数做初始化？
    - Low-level API方法：当前在models里的一些例子，采用如下方法，必需定义[`predicate`](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/io.py#L168)的一个函数传入给`load_vars`才能只load部分参数。
    ```python
    place = fluid.CUDAPlace(0)  # or fluid.CPUPlace()
    exe = fluid.Executor(place)
    exe.run(fluid.default_startup_program())
    def if_exist(var):
        return os.path.exists(os.path.join(pretrained_model_dir, var.name))
    fluid.io.load_vars(exe, pretrained_model_dir, predicate=if_exist)
    ```
    - 问题：High level API如何做？
[io.load_persistables(exe, dirname=param_path)](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/trainer.py#L142)当前这样使用，无法只load部分参数。

3.  如何做内存优化？
    - Low-level API目前使用如下：
       ```
        # 定义网络部分省略
        opts = optimizer.minimize(avg_cost)
        fluid.memory_optimize(fluid.default_main_program()) #优化内存
        ```
    - 问题：High level API如何做？

**Comment by dzhwinter** [comment]
I can answer the question about memory optimize, we will apply the memory optimize to program by default, seems we didn't have the needs to export a high-level API.

**Comment by BigFishMaster** [comment]
@dzhwinter opening the memory optimization will cause performance problem, which have not been solved. Thus, it is necessary to export an interface and set it False.

**Comment by BigFishMaster** [comment]
4. 如何进行每隔一定轮数的训练信息输出？
5.给定demo中的train_network()函数是无入口参数的，真实用的很多情况需要支持入口参数，这个接口不太灵活。
`
def train_network():
    predict = inference_network()
    label = fluid.layers.data(name='label', shape=[1], dtype='int64')
    cost = fluid.layers.cross_entropy(input=predict, label=label)
    avg_cost = fluid.layers.mean(cost)
    accuracy = fluid.layers.accuracy(input=predict, label=label)
    return [avg_cost, accuracy]
`

**Comment by jacquesqiao** [comment]
> 给定demo中的train_network()函数是无入口参数的，真实用的很多情况需要支持入口参数(如class_number等)，这个接口不灵活。

函数参数的问题，这个可以通过python的partial function来解决

**Comment by lgone2000** [comment]
隐藏训练迭代部分 不是通常框架的方式，建议还是要暴露出训练feed数据的主循环。另一方面是方便 写调试代码。目前event回调方式很不灵活，只能在一个全局函数里打打日志 干不了其他的。

**Comment by shanyi15** [comment]
您好，此issue在近一个月内暂无更新，我们将于今天内关闭。若在关闭后您仍需跟进提问，可重新开启此问题，我们将在24小时内回复您。因关闭带来的不便我们深表歉意，请您谅解~感谢您对PaddlePaddle的支持! 
 Hello, this issue has not been updated in the past month. We will close it today for the sake of other user‘s experience. If you still need to follow up on this question after closing, please feel free to reopen it. In that case, we will get back to you within 24 hours. We apologize for the inconvenience caused by the closure and thank you so much for your support of PaddlePaddle Group!
